ocaml/xenopsd/xc/device.ml

```diff
@@ -3828,20 +3828,21 @@ module Dm = struct
                 )
               in
               if not Service.Qemu.(SignalMask.has signal_mask domid) then
-                match Service.Qemu.pid ~xs domid with
-                | None ->
-                    (* after expected qemu stop or domain xs tree destroyed:
-                       this event arrived too late, nothing to do *)
-                    debug
-                      "domid=%d qemu-pid=%d: already removed from xenstore \
-                       during domain destroy"
-                      domid (getpid qemu_pid)
-                | Some _ ->
-                    (* before expected qemu stop: qemu-pid is available in
-                       domain xs tree: signal action to take *)
-                    xs.Xs.write
-                      (Service.Qemu.pidxenstore_path_signal domid)
-                      crash_reason
+                try
+                  xs.Xs.read (Service.Qemu.pidxenstore_path domid) |> ignore ;
+                  (* before expected qemu stop: qemu-pid is available in
+                     domain xs tree: signal action to take *)
+                  xs.Xs.write
+                    (Service.Qemu.pidxenstore_path_signal domid)
+                    crash_reason
+                with _ ->
+                  (* read Service.Qemu.pidxenstore_path from xs failed, should
+                     be after expected qemu stop or domain xs tree destroyed:
+                     this event arrived too late, nothing to do *)
+                  debug
+                    "domid=%d qemu-pid=%d: already removed from xenstore \
+                     during domain destroy"
+                    domid (getpid qemu_pid)
```

Comment:
@psafont Would you please help to review this change, thank you!

Reply:
Yes, this code in device.ml should not need any details about whether the pid in read from a file, xenstore, or both. The current code in the exception handler writes to a xenstore path related to the pid, which is already dubious (and I've heard it doesn't quite work anyway).

Reply:
Hi @psafont , thanks for the review!
The change here is made is not because the main behavior of `pid` is changed: the pid is still get from pidfile and from xenstore as a fallback.
This change is made as some [validation](https://github.com/gangj/xen-api/blob/private/gangj/CA-382035/ocaml/xenopsd/xc/service.ml#L307-L311) for the `pid` is done before returning from the `pid` function in this PR, as this [comment](https://github.com/xapi-project/xen-api/pull/5325#issuecomment-1886514599).

Reply:
I'm keen to merge the main fix in service.ml soon, and separate think about how to improve this on-crash behaviour.

Reply:
I think the purpose of the [pid call](https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/device.ml#L3820) here is to check if the qemu-pid is still available in xs as:
1) if qemu-dm-<domid> is [killed by xapi](https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L391), the crash handler should not be triggered, it is implemented by [clearing the `qemu-pid` immediately after killing and surrounded by a signal_mask](https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L387-L397). I think the surrounding `signal_mask` is to avoid an extreme condition: just after the [killing](https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L391) and before the [xs rm](https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L393), the crash handler is triggered, then it can not distinguish if the killing is done by xapi or not.
2) I think as `qemu-pid` will only be cleared by xapi: [clearing after killing](https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L393) or cleared when the [while dom_path is cleared](https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/domain.ml#L738-L742). it is used to check if the qemu process is stopped by xapi, if yes, no crash handling is needed, otherwise, we need the crash handler to do some cleanup.

Reply:
> I'm keen to merge the main fix in service.ml soon, and separately think about how to improve this on-crash behaviour.

OK, let me add some more logs finally, should not take much time.

Reply:
Log added.